In [1]:
# Outside imports
import os
import importlib
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences

Using TensorFlow backend.


In [22]:
import model
import train
import evaluate
import train_util
import data_util.data
import data_util.batcher
import data_util.config
import data_util.preprocess

importlib.reload(data_util.preprocess)
importlib.reload(train)
importlib.reload(model)
importlib.reload(evaluate)
importlib.reload(train_util)
importlib.reload(data_util.config)
importlib.reload(data_util.data)
importlib.reload(data_util.batcher)

from train import *
from evaluate import *
from model import *
from train_util import *
from data_util.data import *
from data_util.batcher import *
from data_util.preprocess import *

In [5]:
config.sum_path

'/Users/rowancassius/Desktop/capstone/LSTM_Summarizer/data/sum'

In [6]:
task_data = pd.read_csv(os.path.join(config.sum_path, 'task_train_processed.tsv'))
wiki_

In [9]:
# Process the the data
def prep_data(df):
    df['Context'] = df['Context'].map(lambda x: article_process_text(x))
    df['TaskSentence'] = df['TaskSentence'].map(lambda x: article_process_text(x))
    df['Summary'] = df['Summary'].map(lambda x: summary_process_text(x))
    return df
train_data = prep_data(train_data)
# test_data = prep_data(test_data)

In [7]:
vocab = Vocab(os.path.join(config.vocab_path, 'vocab3.txt'))

Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated word in vocabulary file: "
Duplicated w

In [29]:
opt = Namespace(
    train_mle = "yes", 
    train_rl = "no", 
    mle_weight = 1., 
    load_model = None,
    new_lr = None, 
    rl_weight = 0.)

wikitask_batcher = TaskBatcher(
    examples=train_data.to_dict('records'),
    vocab=vocab,
    mode='train',
    batch_size=32,
    single_pass=False
)

# val_task_batcher = TaskBatcher( # Batching obj
#     examples=test_data.to_dict('records')[:200],
#     vocab=vocab, 
#     mode='train', 
#     batch_size=50, 
#     single_pass=False
# )

# train_processor = TaskTrain(vocab, task_batcher, opt, TaskModel, val_task_batcher)
train_processor = TaskTrain(vocab, task_batcher, opt, TaskModel)

# load pre-trained embedding weights
train_processor.model.load_embeddings("embedding_43861_200.tar")

In [30]:
# train
mle_losses = train_processor.trainIters(n_iters=20, report_every=1, save_every = 5)

iter: 1 mle_loss: 6.868 mle_loss_val: -100.0000
iter: 2 mle_loss: 6.675 mle_loss_val: -100.0000
iter: 3 mle_loss: 7.420 mle_loss_val: -100.0000
iter: 4 mle_loss: 6.644 mle_loss_val: -100.0000
iter: 5 mle_loss: 6.621 mle_loss_val: -100.0000
model saved at: 
 data/saved_models/0000005.tar
iter: 6 mle_loss: 5.994 mle_loss_val: -100.0000
iter: 7 mle_loss: 5.094 mle_loss_val: -100.0000
iter: 8 mle_loss: 4.194 mle_loss_val: -100.0000
iter: 9 mle_loss: 4.651 mle_loss_val: -100.0000
iter: 10 mle_loss: 3.671 mle_loss_val: -100.0000
model saved at: 
 data/saved_models/0000010.tar
iter: 11 mle_loss: 3.678 mle_loss_val: -100.0000
iter: 12 mle_loss: 4.053 mle_loss_val: -100.0000
-------------------Keyboard Interrupt------------------


NameError: name 'exit' is not defined

In [24]:
mle_losses

{'train': [(1, 6.6681809425354),
  (2, 7.459186553955078),
  (3, 6.613460540771484),
  (4, 6.715197563171387),
  (5, 5.890581130981445),
  (6, 5.120395660400391)],
 'val': []}

In [32]:
opt_retrain = Namespace(
    train_mle = "yes", 
    train_rl = "no", 
    mle_weight = 1., 
    load_model = '0000010.tar',
    new_lr = None, 
    rl_weight = 0.)

train_processor = TaskTrain(vocab, task_batcher, opt_retrain, TaskModel)

mle_losses = train_processor.trainIters(n_iters=5, report_every=1, save_every = 5)

Loaded model at data/saved_models/0000010.tar
iter: 1 mle_loss: 3.565 mle_loss_val: -100.0000
-------------------Keyboard Interrupt------------------


NameError: name 'exit' is not defined

In [ ]:
mle_losses

Decoding Time

In [28]:
len(train_data.to_dict('records') + test_data.to_dict('records'))

2820

In [26]:
dat.shape

(2820, 11)

In [29]:
load_model = os.path.join(config.log_root, "data/saved_models_2/0000360.tar") # model directory

# new batcher for evaluation
task_batcher = TaskBatcher( # Batching obj
    examples=train_data.to_dict('records') + test_data.to_dict('records'),
    vocab=vocab, 
    mode='train', 
    batch_size=282,    # 188, 
    single_pass=True)

eval_processor = TaskEvaluate(vocab, task_batcher, TaskModel, load_model) # Evaluation object

In [30]:
2256/8

282.0

In [31]:
train_data.shape

(2256, 11)

In [32]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

In [33]:
count_parameters(eval_processor.model)

14911184

In [34]:
decoded_sents, ref_sents, task_sents, context_sents = eval_processor.evaluate_batch()

Summarizing Batch...
Summarizing Batch...
example_generator completed reading all examples. No more data.


Exception in thread Thread-9:
Traceback (most recent call last):
  File "/Users/rowancassius/Desktop/capstone/LSTM_Summarizer/data_util/batcher.py", line 448, in text_generator
    example = next(example_generator)
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rowancassius/opt/anaconda3/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/Users/rowancassius/opt/anaconda3/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/rowancassius/Desktop/capstone/LSTM_Summarizer/data_util/batcher.py", line 429, in fill_example_queue
    context, task, summary = next(input_gen) # read the next example from file. article and abstract are both strings.
RuntimeError: generator raised StopIteration



Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
Summarizing Batch...
INFO:tensorflow:Finished reading dataset in single_pass mode.


In [35]:
len(ref_sents)

2820

In [36]:
len(decoded_sents)

2820

In [37]:
scores = Rouge().get_scores(decoded_sents, ref_sents, avg = True)

In [38]:
scores

{'rouge-1': {'f': 0.6020702789916959,
  'p': 0.7667154319281988,
  'r': 0.5381503106904498},
 'rouge-2': {'f': 0.3904763855618124,
  'p': 0.5170605369807497,
  'r': 0.35230873911421756},
 'rouge-l': {'f': 0.5992272203650424,
  'p': 0.7633533920767971,
  'r': 0.5348177231306076}}

In [22]:
pd.set_option('display.max_colwidth', -1)

In [42]:
df = pd.DataFrame({'task': task_sents,'ref': ref_sents, 'decoded': decoded_sents})
df.sample(10)

task  \
2457  Please send a package of article reprints to the attention of Joan Payne at the following address:                                                                                         
2536  Please invite Steve Schwarzbach                                                                                                                                                            
2751  Check out the E-Flyer below and let me know what's up.                                                                                                                                     
2683  Please remove the Personal Expense icon from all personal expense entries and submit the expense report once again for payment.                                                            
1843  As soon as you get confirmation that your request has been completed, please let Edmund know so he can update/replace the London documents in the Financial Trading Agreements Database.   
1391  Please take me off of this distribution.                                                                                                                                                   
1047  Tom: Please forward your requisite email request to Donna Lowry, Rick Carson and Bill Bradford with your request and reasons.                                                              
2025  To register for the Energy Town Hall in Dallas on Wednesday, Nov.28, please simply reply to this email with your contact information.                                                      
536   Please take a look at this email ....                                                                                                                                                      
2533  please call after reading                                                                                                                                                                  

                                                                                   ref  \
2457  send package of article reprints to Joan Payne                                     
2536  invite Steve Schwarzbach                                                           
2751  look at the e - flyer                                                              
2683  remove personal expense icon from entries and resubmit expense report              
1843  confirm request completion with Edmund                                             
1391  remove SENDER from distribution                                                    
1047  forward requisite email request to Donna Lowry , Rick Carson , and Bill Bradford   
2025  reply to email with contact information to register for energy town hall           
536   examine this email                                                                 
2533  call SENDER                                                                        

                              decoded  
2457  send package of article          
2536  invite Steve Schwarzbach         
2751  check out e - flyer              
2683  remove personal expense          
1843  update SENDER                    
1391  take SENDER off of distribution  
1047  forward requisite request        
2025  reply to SENDER                  
536   look at email                    
2533  call SENDER

In [43]:
df.to_csv(os.path.join(config.log_root, 'data/test_results_2.csv'), sep = '\t', header=True, index=False)

In [44]:
'packet' in vocab._word_to_id

True

In [45]:
'interview' in vocab._word_to_id

True

Checking certain words for presence in the vocab:

In [46]:
words = ["ensure", "indicate", "turn", "open", "add"]

In [47]:
[w in vocab._word_to_id for w in words]

[True, True, True, True, True]

In [48]:
action_words = train_data['Summary'].map(lambda x: x.split(' ', 1)[0])

In [49]:
action_words.dtype

dtype('O')

In [50]:
in_vocab = action_words.map(lambda w: w in vocab._word_to_id)

In [51]:
len(in_vocab)

2256

In [52]:
in_vocab.sum()

2240

In [53]:
action_words[~in_vocab]

1354    revier  
76      relpy   
2190    briing  
1213    exame   
2192    email/  
126     reiview 
990     hande   
1366    foraard 
1988    confirm/
14      reissue 
25      exectute
2627    meesage 
1425    complute
416     pring   
1914    verity  
492     helo    
Name: Summary, dtype: object

In [54]:
test_data[~in_vocab]

IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).

2240 of 2256 verbs to start summaries are in the vocab

In [ ]:
# train
action_words[~in_vocab]

In [ ]:
train_data[~in_vocab]['Labeler'].value_counts()

In [ ]:
train_data[~in_vocab]['Summary']

Look at Worst summaries

In [55]:
def score(r):
    scores = Rouge().get_scores(r.decoded, r.ref)[0]
    return [scores['rouge-1']['f'], scores['rouge-2']['f'], scores['rouge-l']['f']]

In [56]:
scored_examples = [score(r) for r in df.itertuples()]

In [57]:
df = pd.concat([df, pd.DataFrame(scored_examples, columns=['Rouge-1', 'Rouge-2', 'Rouge-L'])], axis=1)

In [58]:
df

task  \
0     If you would like to take advantage of this, please fax a completed registration form, along with a brief covering note referring to this email, and I will ensure that our administration depart handle the relevant paper work.   
1     If you are within 12 months of graduating and haven't already sent me an email, please reply letting me know you are eligible and interested in participating in registering for the fall 01 - spring 02 OCR program.?              
2     Please reallocate mtr 1517 with these changes; sale of 10,000 on deal 93804, purchase of 4,000 on 148851.                                                                                                                           
3     If you are not an intended recipient of this message, please do not read, copy or disclose the contents of this communication to others and please notify the sender at 713-2077-414 .                                              
4     Could you please take a look at the document it's a very quick read , call me if you need more information, and then vote for which option you would prefer?                                                                        
...                                                                                                                                                            ...                                                                        
2815  Can you call me?                                                                                                                                                                                                                    
2816  Please respond to EnergyWriter@aol.com                                                                                                                                                                                              
2817  Please respond to loughrid@rice.edu                                                                                                                                                                                                 
2818  Please respond to Whitepapers                                                                                                                                                                                                       
2819  please print out                                                                                                                                                                                                                    

                                            ref                      decoded  \
0     fax complete registration form to SENDER   fax completed form            
1     reply to SENDER about OCR program          reply to SENDER               
2     reallocate mtr 1517                        reallocate with changes       
3     do not read contents of this commuication  notify read at 713-2077-414   
4     read document and vote                     call SENDER                   
...                      ...                             ...                   
2815  call SENDER                                call SENDER                   
2816  respond to energywriter@aolcom             respond to [UNK]              
2817  respond to loughrid@riceedu                respond to [UNK]              
2818  respond to whitepapers                     respond to whitepapers        
2819  print                                      print out out                 

       Rouge-1   Rouge-2   Rouge-L  
0     0.444444  0.000000  0.444444  
1     0.666667  0.571429  0.666667  
2     0.333333  0.000000  0.333333  
3     0.181818  0.000000  0.181818  
4     0.000000  0.000000  0.000000  
...        ...       ...       ...  
2815  1.000000  1.000000  1.000000  
2816  0.666667  0.500000  0.666667  
2817  0.666667  0.500000  0.666667  
2818  1.000000  1.000000  1.000000  
2819  0.500000  0.000000  0.666667  



In [62]:
df_bad = df[df['Rouge-1'] < 0.1]

In [65]:
df_bad

task  \
4     Could you please take a look at the document it's a very quick read , call me if you need more information, and then vote for which option you would prefer?                                           
30    Please take a few minutes today to provide much needed and confidential comments regarding Enron and your work environment.                                                                            
76    Peter Feltman/ENRON@enronXgate 04/10/2001 02:31 PM To: Mark Greenberg/NA/Enron@ENRON cc: Mathis Conner/ENRON@enronXgate Subject: Global Crossing's comments to the Enron NDA - Please responed ASAP!   
78    Can you pass up this opportunity??                                                                                                                                                                     
85    Please look over it carefully and if there is language you are going to provide please do it post haste.                                                                                               
...                                                                                                        ...                                                                                               
2702  -bcc: X-Folder: Tanya Jones Dec2000 Notes Folders All documents X-Origin: JONES-T X-FileName: tjones.nsf                                                                                               
2704  After you read through it, please get back with me.                                                                                                                                                    
2766  I assume that this request relates to corporate functions under Rick.                                                                                                                                  
2778  Can you see if this is the correct CP.                                                                                                                                                                 
2810  -Origin: Taylor-M X-FileName: mtaylor.nsf                                                                                                                                                              

                                        ref                        decoded  \
4     read document and vote                 call SENDER                     
30    comment on work conditions             provide today to confidential   
76    respond to SENDER                      [UNK]                           
78    get back to SENDER                     pass up                         
85    review it and respond with language    look post                       
...                                   ...          ...                       
2702  send to SENDER after recieving         t notes                         
2704  respond to SENDER after reading email  read back                       
2766  submit by tomorrow                     corporate request               
2778  check this for correctness             review correct                  
2810  contact Fred Lagrasta                  Taylor-M X-FileName             

      Rouge-1  Rouge-2  Rouge-L  
4     0.0      0.0      0.0      
30    0.0      0.0      0.0      
76    0.0      0.0      0.0      
78    0.0      0.0      0.0      
85    0.0      0.0      0.0      
...   ...      ...      ...      
2702  0.0      0.0      0.0      
2704  0.0      0.0      0.0      
2766  0.0      0.0      0.0      
2778  0.0      0.0      0.0      
2810  0.0      0.0      0.0      

[143 rows x 6 columns]

In [64]:
import urllib.request

In [ ]:
urllib.request.urlretrieve()